### PR TITLE
refactor(di): unified di injector and core injector

### PR DESCRIPTION
**Pull request by vsavkin**
[Review on Reviewable](https://reviewable.io/reviews/angular/angular/2737)

### CODE REVIEW

File: modules/angular2/src/core/compiler/element_injector.ts [hunk 1]
```diff
@@ -665,13 +444,21 @@ export class ElementInjector extends TreeNode<ElementInjector> {
 
   hydrated: boolean;
 
-  _strategy: _ElementInjectorStrategy;
+  _injector: Injector;
```
Comment (vsavkin) [inline modules/angular2/src/core/compiler/element_injector.ts]: should be private?

File: modules/angular2/src/core/compiler/element_injector.ts [hunk 2]
```diff
@@ -665,13 +444,21 @@ export class ElementInjector extends TreeNode<ElementInjector> {
...
 
   constructor(public _proto: ProtoElementInjector, parent: ElementInjector) {
     super(parent);
-    this._strategy = _proto._strategy.createElementInjectorStrategy(this);
 
-    this._constructionCounter = 0;
+    this._injector = new Injector(this._proto.protoInjector);
+
+    var injectorStrategy = <any>this._injector.strategy;
+    this._strategy = injectorStrategy instanceof InjectorInlineStrategy ?
```
Comment (vsavkin) [inline modules/angular2/src/core/compiler/element_injector.ts]: Add a comment explaining why we are coupling ourselves to the strategy type

File: modules/angular2/src/core/compiler/element_injector.ts [hunk 3]
```diff
@@ -1394,106 +941,53 @@ class ElementInjectorInlineStrategy implements _ElementInjectorStrategy {
...
-  constructor(index) {
-    super();
-    this.message = `Index ${index} is out-of-bounds.`;
-  }
-
-  toString(): string { return this.message; }
 }
 
-class QueryError extends BaseException {
+export class QueryError extends BaseException {
```
Comment (vsavkin) [inline modules/angular2/src/core/compiler/element_injector.ts]: Try to remove export

File: modules/angular2/src/di/annotations_impl.ts [hunk 4]
```diff
@@ -124,4 +86,219 @@ export class DependencyAnnotation {
...
+
+  get includeSelf(): boolean { return isBlank(this._includeSelf) ? false : this._includeSelf; }
+
+  toString() {
+    return `@Visibility(depth: ${this.depth}, crossComponentBoundaries: ${this.crossComponentBoundaries}, includeSelf: ${this.includeSelf}})`;
+  }
+}
+
+/**
+ * Specifies that an injector should retrieve a dependency from its element.
```
Comment (vsavkin) [inline modules/angular2/src/di/annotations_impl.ts]: Fix the docs

File: modules/angular2/src/di/annotations_impl.ts [hunk 5]
```diff
@@ -124,4 +86,219 @@ export class DependencyAnnotation {
...
+ *
+ * @exportedAs angular2/di
+ */
+@CONST()
+export class Self extends Visibility {
+  constructor() { super(0, false, true); }
+  toString() { return `@Self()`; }
+}
+
+// make constants after switching to ts2dart
```
Comment (vsavkin) [inline modules/angular2/src/di/annotations_impl.ts]: Remove the comment

File: modules/angular2/src/di/binding.ts [hunk 6]
```diff
@@ -508,12 +438,18 @@ function _extractToken(typeOrFunc, annotations /*List<any> | any*/,
...
+    return _createDependency(token, optional, visibility, depProps);
   } else {
     throw new NoAnnotationError(typeOrFunc, params);
   }
 }
 
-function _createDependency(token, asPromise, lazy, optional, depProps): Dependency {
-  return new Dependency(Key.get(token), asPromise, lazy, optional, depProps);
+function _defaulVisiblity(typeOrFunc) {
+  if (!(typeOrFunc instanceof Type)) return unbounded;
```
Comment (vsavkin) [inline modules/angular2/src/di/binding.ts]: Check if I have reflection capabilities and do not, then default to unbounded? Or maybe always require visibility.

File: modules/angular2/src/di/exceptions.ts [hunk 7]
```diff
@@ -198,3 +198,13 @@ export class NoAnnotationError extends BaseException {
 
   toString(): string { return this.message; }
 }
+
+export class OutOfBoundsError extends BaseException {
```
Comment (vsavkin) [inline modules/angular2/src/di/exceptions.ts]: Add a comment

File: modules/angular2/src/di/injector.ts [hunk 8]
```diff
@@ -8,22 +8,373 @@ import {
...
+  visibility2: number = null;
+  visibility3: number = null;
+  visibility4: number = null;
+  visibility5: number = null;
+  visibility6: number = null;
+  visibility7: number = null;
+  visibility8: number = null;
+  visibility9: number = null;
+
+  constructor(protoEI: ProtoInjector, bd: List<any>) {
```
Comment (vsavkin) [inline modules/angular2/src/di/injector.ts]: Type bd

File: modules/angular2/src/di/injector.ts [hunk 9]
```diff
@@ -8,22 +8,373 @@ import {
...
+
+  createInjectorStrategy(injector: Injector): InjectorStrategy {
+    return new InjectorInlineStrategy(injector, this);
+  }
 }
 
-function _isWaiting(obj): boolean {
-  return obj instanceof _Waiting;
+export class ProtoInjectorDynamicStrategy implements ProtoInjectorStrategy {
+  bindings: List<ResolvedBinding>;
```
Comment (vsavkin) [inline modules/angular2/src/di/injector.ts]: Change all lists to ResolvedBinding[].

File: modules/angular2/src/router/router_outlet.ts [hunk 10]
```diff
@@ -26,15 +26,17 @@ export class RouterOutlet {
...
   constructor(elementRef: ElementRef, private _loader: DynamicComponentLoader,
-              private _parentRouter: routerMod.Router, private _injector: Injector,
+              private _parentRouter: routerMod.Router, _injector: Injector,
               @Attribute('name') nameAttr: string) {
     // TODO: reintroduce with new // sibling routes
     // if (isBlank(nameAttr)) {
     //  nameAttr = 'default';
     //}
 
+    this._injector = _injector.getAppInjector();
```
Comment (vsavkin) [inline modules/angular2/src/router/router_outlet.ts]: Remove it

File: modules/angular2/src/test_lib/perf_util.ts [hunk 11]
```diff
@@ -1,6 +1,11 @@
 export {verifyNoBrowserErrors} from './e2e_util';
 
-import {bind, Options} from 'benchpress/benchpress';
+// import {bind, Options} from 'benchpress/benchpress';
```
Comment (vsavkin) [inline modules/angular2/src/test_lib/perf_util.ts]: Uncomment it

File: modules/angular2/test/core/application_spec.ts [hunk 12]
```diff
@@ -108,24 +109,6 @@ export function main() {
       expect(refPromise).not.toBe(null);
     });
 
-    it('should resolve an injector promise and contain bindings',
```
Comment (vsavkin) [inline modules/angular2/test/core/application_spec.ts]: Bring one of them back

File: modules/angular2/test/core/compiler/integration_dart_spec.dart [hunk 13]
```diff
@@ -55,7 +55,8 @@ main() {
   });
 
   describe('Error handling', () {
-    it('should preserve Error stack traces thrown from components', inject([
+    //TODO: vsavkin reenable this test after merging DI and EI
+    xit('should preserve Error stack traces thrown from components', inject([
```
Comment (vsavkin) [inline modules/angular2/test/core/compiler/integration_dart_spec.dart]: Talk to Jbanov

File: modules/angular2/test/di/injector_spec.ts [hunk 14]
```diff
@@ -396,11 +342,25 @@ export function main() {
...
         ]);
-        var binding = bindings[Key.get("token").id];
+        var binding = bindings[0];
 
         expect(binding.dependencies[0].key.token).toEqual("dep");
         expect(binding.dependencies[0].properties).toEqual([new CustomDependencyAnnotation()]);
       });
     });
+
+    describe("default visibility", () => {
```
Comment (vsavkin) [inline modules/angular2/test/di/injector_spec.ts]: Add a test for the case when reflector is not provided.

File: modules/benchpress/src/runner.ts [hunk 15]
```diff
@@ -49,9 +49,9 @@ export class Runner {
     if (isPresent(bindings)) {
       sampleBindings.push(bindings);
     }
-    return Injector.resolveAndCreate(sampleBindings)
-        .asyncGet(Sampler)
-        .then((sampler) => sampler.sample());
+
+    var sampler = PromiseWrapper.resolve(Injector.resolveAndCreate(sampleBindings).get(Sampler));
```
Comment (vsavkin) [inline modules/benchpress/src/runner.ts]: Check if I have to do it

File: tsconfig.json [hunk 16]
```diff
@@ -11,6 +11,8 @@
     },
     "filesGlob": [
         "./**/*.ts",
-        "!./node_modules/**/*.ts"
+        "!./node_modules/**/*.ts",
```
Comment (vsavkin) [inline tsconfig.json]: Fix tsconfig.json

**Comment by mhevery** [comment]
Review status: 0 of 51 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

<sup>**[modules/angular2/angular2.api.ts, line 25 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2737#-Jsh0MGx-ue_khajbGMV)** ([raw file](https://github.com/angular/angular/blob/9bd3c81c4566bbb97d7b7f18193497b451312ab8/modules/angular2/angular2.api.ts#L25)):</sup>
Here is a task for you.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/angular/angular/2737)

**Comment by vsavkin** [comment]
Review status: 0 of 51 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

<sup>**[modules/angular2/angular2.api.ts, line 25 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2737#-Jsh0MGx-ue_khajbGMV)** ([raw file](https://github.com/angular/angular/blob/9bd3c81c4566bbb97d7b7f18193497b451312ab8/modules/angular2/angular2.api.ts#L25)):</sup>
Done.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/angular/angular/2737)

**Comment by mhevery** [comment]
<sup>**[modules/angular2/angular2.api.ts, line 25 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2737#-Jsh0MGx-ue_khajbGMV)** ([raw file](https://github.com/angular/angular/blob/9bd3c81c4566bbb97d7b7f18193497b451312ab8/modules/angular2/angular2.api.ts#L25)):</sup>
Love it!

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/angular/angular/2737)

**Comment by mhevery** [comment]
Reviewed 51 of 51 files at r1.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

<sup>**[modules/angular2/src/core/application.ts, line 93 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2737#-Jsh1U0FD9SMhr3yEgr8)** ([raw file](https://github.com/angular/angular/blob/9bd3c81c4566bbb97d7b7f18193497b451312ab8/modules/angular2/src/core/application.ts#L93)):</sup>
add type if relevant?

---

<sup>**[modules/angular2/src/di/injector.ts, line 609 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2737#-Jsh2baXFjZUbD6Pc9QO)** ([raw file](https://github.com/angular/angular/blob/9bd3c81c4566bbb97d7b7f18193497b451312ab8/modules/angular2/src/di/injector.ts#L609)):</sup>
GT needs more then 10, can we do 20? We also used to have a test for this. See: https://github.com/angular/angular/commit/5ab92ce8440927b302f79444d76eae26863f9b52

---

<sup>**[modules/benchpress/src/web_driver_extension.ts, line 25 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2737#-Jsh4vGZ0Ne5qmPnt_Fe)** ([raw file](https://github.com/angular/angular/blob/9bd3c81c4566bbb97d7b7f18193497b451312ab8/modules/benchpress/src/web_driver_extension.ts#L25)):</sup>
add type annotation please

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/angular/angular/2737)

File: modules/angular2/di.ts [hunk 17]
```diff
@@ -7,16 +7,39 @@
...
+  Unbounded
+} from './src/di/decorators';
+export {self} from './src/di/annotations_impl';
 export {forwardRef, resolveForwardRef, ForwardRefFn} from './src/di/forward_ref';
-export {resolveBindings, Injector} from './src/di/injector';
+export {
+  resolveBindings,
+  Injector,
+  ProtoInjector,
+  PUBLIC_AND_PRIVATE,
```
Comment (vicb) [inline modules/angular2/di.ts]: nit: `PUBLIC_AND_PRIVATE` sounds a little bit weird

File: modules/angular2/src/di/annotations_impl.ts [hunk 18]
```diff
@@ -124,4 +86,218 @@ export class DependencyAnnotation {
...
+@CONST()
+export class Visibility extends DependencyAnnotation {
+  constructor(public depth: number, public crossComponentBoundaries: boolean,
+              public _includeSelf: boolean) {
+    super();
+  }
+
+  get includeSelf(): boolean { return isBlank(this._includeSelf) ? false : this._includeSelf; }
+
+  toString() {
```
Comment (vicb) [inline modules/angular2/src/di/annotations_impl.ts]: return type

File: modules/angular2/src/di/annotations_impl.ts [hunk 19]
```diff
@@ -124,4 +86,218 @@ export class DependencyAnnotation {
...
+ * ```
+ *<div dependency="1" my-directive></div>
+ * ```
+ *
+ * @exportedAs angular2/di
+ */
+@CONST()
+export class Self extends Visibility {
+  constructor() { super(0, false, true); }
+  toString() { return `@Self()`; }
```
Comment (vicb) [inline modules/angular2/src/di/annotations_impl.ts]: return type

File: modules/angular2/src/di/injector.ts [hunk 20]
```diff
@@ -8,22 +8,371 @@ import {
...
 import {resolveForwardRef} from './forward_ref';
+import {self, unbounded} from './annotations_impl';
 
 const _constructing = CONST_EXPR(new Object());
 const _notFound = CONST_EXPR(new Object());
 
-class _Waiting {
-  constructor(public promise: Promise<any>) {}
+// Threshold for the dynamic version
+var _MAX_CONSTRUCTION_COUNTER = 10;
```
Comment (vicb) [inline modules/angular2/src/di/injector.ts]: const ?

File: modules/angular2/src/facade/async.ts [hunk 21]
```diff
@@ -28,6 +28,16 @@ export class PromiseWrapper {
     return promise.then(success, rejection);
   }
 
+  static wrap<T>(computation: () => any): Promise<T> {
```
Comment (vicb) [inline modules/angular2/src/facade/async.ts]: `computation: () => T` ?

**Comment by vicb** [comment]
`.asyncGet` got dropped, should it be mentioned in the commit message (BREAKING CHANGE) - any other BC worth mentionning.

Also should this close #1111, #2407, #2408 ?

**Comment by PascalPrecht** [comment]
Also, it seems like a couple of decorators have been removed which should also be mentioned.

**Comment by vsavkin** [comment]
Reviewed 1 of 51 files at r1.
Review status: 41 of 51 files reviewed at latest revision, 23 unresolved discussions, some commit checks pending.

---

<sup>**[modules/angular2/di.ts, line 36 \[r2\]](https://reviewable.io:443/reviews/angular/angular/2737#-JsgWQgPa-Io2KKfDM0T-r2-36)** ([raw file](https://github.com/angular/angular/blob/4e8b04b939a5a331b4bd8bf053156aba74320762/modules/angular2/di.ts#L36)):</sup>
Sure, we can rename it.

---

<sup>**[modules/angular2/src/core/application.ts, line 93 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2737#-Jsh1U0FD9SMhr3yEgr8)** ([raw file](https://github.com/angular/angular/blob/9bd3c81c4566bbb97d7b7f18193497b451312ab8/modules/angular2/src/core/application.ts#L93)):</sup>
Done.

---

<sup>**[modules/angular2/src/di/annotations_impl.ts, line 101 \[r2\]](https://reviewable.io:443/reviews/angular/angular/2737#-JsgWQgQnkroESeULyf3-r2-101)** ([raw file](https://github.com/angular/angular/blob/4e8b04b939a5a331b4bd8bf053156aba74320762/modules/angular2/src/di/annotations_impl.ts#L101)):</sup>
Done.

---

<sup>**[modules/angular2/src/di/annotations_impl.ts, line 107 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2737#-JsgWQgQnkroESeULyf3-r1-107)** ([raw file](https://github.com/angular/angular/blob/9bd3c81c4566bbb97d7b7f18193497b451312ab8/modules/angular2/src/di/annotations_impl.ts#L107)):</sup>
Done.

---

<sup>**[modules/angular2/src/di/annotations_impl.ts, line 146 \[r2\]](https://reviewable.io:443/reviews/angular/angular/2737#-JsgWQgQnkroESeULyf3-r2-146)** ([raw file](https://github.com/angular/angular/blob/4e8b04b939a5a331b4bd8bf053156aba74320762/modules/angular2/src/di/annotations_impl.ts#L146)):</sup>
Done.

---

<sup>**[modules/angular2/src/di/injector.ts, line 23 \[r2\]](https://reviewable.io:443/reviews/angular/angular/2737#-JsgWQgQnkroESeULyf7-r2-23)** ([raw file](https://github.com/angular/angular/blob/4e8b04b939a5a331b4bd8bf053156aba74320762/modules/angular2/src/di/injector.ts#L23)):</sup>
Done.

---

<sup>**[modules/angular2/src/di/injector.ts, line 145 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2737#-JsgWQgQnkroESeULyf7-r1-147)** ([raw file](https://github.com/angular/angular/blob/9bd3c81c4566bbb97d7b7f18193497b451312ab8/modules/angular2/src/di/injector.ts#L145)):</sup>
Done.

---

<sup>**[modules/angular2/src/di/injector.ts, line 607 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2737#-Jsh2baXFjZUbD6Pc9QO)** ([raw file](https://github.com/angular/angular/blob/9bd3c81c4566bbb97d7b7f18193497b451312ab8/modules/angular2/src/di/injector.ts#L607)):</sup>
Done.

---

<sup>**[modules/angular2/src/facade/async.ts, line 31 \[r2\]](https://reviewable.io:443/reviews/angular/angular/2737#-JsgWQgQnkroESeULyfA-r2-31)** ([raw file](https://github.com/angular/angular/blob/4e8b04b939a5a331b4bd8bf053156aba74320762/modules/angular2/src/facade/async.ts#L31)):</sup>
Done.

---

<sup>**[modules/benchpress/src/web_driver_extension.ts, line 25 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2737#-Jsh4vGZ0Ne5qmPnt_Fe)** ([raw file](https://github.com/angular/angular/blob/9bd3c81c4566bbb97d7b7f18193497b451312ab8/modules/benchpress/src/web_driver_extension.ts#L25)):</sup>
Done.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/angular/angular/2737)

**Pull request merged**